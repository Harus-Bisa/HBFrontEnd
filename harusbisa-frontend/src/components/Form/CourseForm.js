import React from "react";
import {connect} from "react-redux";
import { addCourse, editCourse } from "../../redux/actions";
import { Form, FormGroup, Label, Input, FormFeedback, FormText, Col} from 'reactstrap';
import { HB_YELLOW } from "../../css/constants/color";
import { Button } from "@material-ui/core";

function mapStateToProps(state, ownProps){
    if (ownProps.id){
        let courses = state.courses
        for (let i=0; i<courses.length; i++){
            if (courses[i]._id === ownProps.id){
                return{
                    course: courses[i]
                }
            }
        }
        return null
    }
    else{
        return {
            course:{
                course_name: "",
                start_term: "",
                end_term: ""
            }
        }
    }
    
}
function CourseForm(props){
    const [name, setName] = React.useState(props.course.course_name);
    const [startDate, setStartDate] = React.useState(props.course.start_term);
    const [endDate, setEndDate] = React.useState(props.course.end_term);
    
    const submit = (event) =>{
        event.preventDefault();
        if (props.id){
            props.editCourse(props.id, name, startDate, endDate);
        }
        else{
            props.addCourse(name, startDate, endDate);
        }
        props.closePopup();
    }
    return(
        <div className="container-fluid">
            {/* <form onSubmit={submit}>
                <label>Course Name*</label>
                <input type="text" value={name} onChange={(event) => setName(event.target.value)}></input>
                <br/>
                <label>Start Date*</label>
                <input type="text" value={startDate} onChange={(event) => setStartDate(event.target.value)}></input>
                <br/>
                <label>End Date*</label>
                <input type="text" value={endDate} onChange={(event) => setEndDate(event.target.value)}></input>
                <br/>
                <button type="submit">Submit</button>
            </form> */}
            <div className="row" style={{borderBottom:'2px solid '+HB_YELLOW}}>
                <div className="col-12" style={{display:'flex', justifyContent:'center', padding:'1rem 2rem'}}>
                    <h3>Tambahkah Kelas Baru</h3>
                </div>
            </div>
            <div className="content">
                <Form onSubmit={submit}>
                    <FormGroup row>
                        <Label sm={5}>Nama Kelas</Label>
                        <Col sm={7}>
                            <Input type="text" value={name} placeholder={"Contoh: Biologi Molekuler Kelas A"} onChange={(event) => setName(event.target.value)}/>
                            <FormText>Nama ini adalah nama yang akan dilihat siswa Anda.</FormText>
                        </Col>
                    </FormGroup>
                    <FormGroup row>
                        <Label sm={5}>Periode Kelas</Label>
                        <Col sm={7}>
                            <div className="row">
                                <div className="col-5">
                                    <Input type="select" name="select" id="exampleSelect">
                                        <option>Juli 2019</option>
                                        <option>Juli 2020</option>
                                        <option>Juli 2021</option>
                                        <option>Juli 2022</option>
                                        <option>Juli 2023</option>
                                    </Input>
                                </div>
                                <div className="col-2">
                                    <p>-</p>
                                </div>
                                <div className="col-5">
                                    <Input type="select" name="select" id="exampleSelect">
                                        <option>Juli 2019</option>
                                        <option>Juli 2020</option>
                                        <option>Juli 2021</option>
                                        <option>Juli 2022</option>
                                        <option>Juli 2023</option>
                                    </Input>
                                </div>
                            </div>
                        </Col>
                    </FormGroup>
                    <Button className="prof-button" type="submit">Submit</Button>
                </Form>
            </div>
        </div>
    )
}

export default connect(mapStateToProps,{addCourse, editCourse})(CourseForm);