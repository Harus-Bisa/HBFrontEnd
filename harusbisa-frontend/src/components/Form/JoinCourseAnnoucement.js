import React from "react";
import { Button } from "@material-ui/core";
import {connect} from "react-redux";

function JoinCourseAnnoucement(props){
    const style={
        text:{
            marginBottom:"1rem"
        }
    }
    return(
        <div className="container-fluid">
            <div className="content">
                <div className="row">
                    <div className="col-12">
                        <h5 style={style.text}>Mintalah siswa Anda membuat akun mereka sendiri</h5>
                    </div>
                </div>
                <div className="row">
                    <div className="col-12">
                        <p style={style.text}>Salin, kemudia bagikan link ini dengan siswa Anda.</p>
                    </div>
                </div>
                <div className="row" style={{marginBottom:'1rem'}}>
                    <div className="col-9" style={{display:'flex'}}>
                        <div style={{border:'1.5px solid #f4f4f4', padding:'4.5px 8px', margin:'auto 0', width:"100%"}}>
                            <p>{"www.harusbisa.net/join/" + props.joinCode}</p>
                        </div>
                    </div>
                    <div className="col-3" style={{display:'flex'}}>
                        <Button className="prof-button" fullWidth style={{margin:'auto'}}>Salin</Button>
                    </div>
                </div>
                <div className="row" style={{marginBottom:'1rem'}}>
                    <div className="col-12">
                        <p>Atau mintalah siswa Anda mengunjungi www.harusbisa.net/signup dan masukkan kode kelas Anda <strong>{props.joinCode}</strong></p>
                    </div>
                </div>
                <div className="row justify-content-end">
                    <div className="col-3">
                        <Button className="prof-button" fullWidth onClick={props.closePopup}>Selesai</Button>
                    </div>
                </div>
                
            </div>
        </div>
    )
}

function MapStateToProps(state){
    return{
        joinCode:state.courses[state.courses.length-1].join_code
    }
}
export default connect(MapStateToProps)(JoinCourseAnnoucement);