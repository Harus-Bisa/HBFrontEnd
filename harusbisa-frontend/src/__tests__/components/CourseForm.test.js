import React from "react";
import {mount} from "enzyme";
import configureMockStore from "redux-mock-store";
import '../../test-config';
import {Provider} from 'react-redux';
import CourseForm from "../../components/Form/CourseForm";
import thunk from 'redux-thunk';
import { Button } from "@material-ui/core";
import {Form} from "reactstrap";
import JoinCourseAnnoucement from "../../components/Form/JoinCourseAnnoucement";
const mockStore = configureMockStore([thunk]);

describe("CourseForm component", () =>{
    let component;
    let store;
    let initialState;
    let mockClosePopup = jest.fn()

    afterEach(() =>{
        jest.clearAllMocks();
    })
    it("matches snapshot", () =>{
        store = mockStore(initialState);
        component = mount(<Provider store={store}><CourseForm/></Provider>)
        expect(component.html()).toMatchSnapshot()
    })
    it("renders edit type form", () =>{
        var id = "5db492c349a67b33b8d0a2a2"
        initialState= {
            courses:[{
                "courseId": "5db492c349a67b33b8d0a2a2",
                "courseName": "Calculus I",
                "courseDescription": "The basic of calculus (derivatives).",
                "startTerm": "August 2019",
                "endTerm": "December 2019",
                "numberOfLectures": 5,
                "numberOfStudents": 21,
                "instructors": [
                "John Doe",
                "Bryan Smith"
                ],
                "joinCode": "123654"
              }]
        }
        store = mockStore(initialState);
        component = mount(<Provider store={store}><CourseForm id={id}/></Provider>)
        var header = component.find("h3")
        expect(header.text()).toBe("Edit Kelas")
    })
    it("renders add type form", () =>{
        store = mockStore(initialState);
        component = mount(<Provider store={store}><CourseForm/></Provider>)
        var header = component.find("h3")
        expect(header.text()).toBe("Tambahkan Kelas Baru")
    })
    it("changes state when form is inputted", () =>{
        let setState = jest.fn();
        let useStateSpy = jest.spyOn(React, 'useState');
        useStateSpy.mockImplementation((init) => [init, setState])
        store = mockStore(initialState);
        component = mount(<Provider store={store}><CourseForm/></Provider>)
        
        var labels = ["courseName", "startDate", "endDate"]
        labels.forEach(l =>{
            var input = component.find("#"+l).at(0)
            expect(setState).not.toHaveBeenCalled()
            input.simulate("change")
            expect(setState).toHaveBeenCalledTimes(1)
            jest.clearAllMocks()

        })
    })
    it("submits edit", () =>{
        var id = "5db492c349a67b33b8d0a2a2"
        initialState= {
            courses:[{
                "courseId": "5db492c349a67b33b8d0a2a2",
                "courseName": "Calculus I",
                "courseDescription": "The basic of calculus (derivatives).",
                "startTerm": "August 2019",
                "endTerm": "December 2019",
                "numberOfLectures": 5,
                "numberOfStudents": 21,
                "instructors": [
                "John Doe",
                "Bryan Smith"
                ],
                "joinCode": "123654"
              }]
        }
        store = mockStore(initialState);
        store.dispatch = jest.fn()
        const mockPreventDefault = jest.fn();
        component = mount(<Provider store={store}><CourseForm id={id} closePopup={mockClosePopup}/></Provider>)

        expect(store.dispatch).not.toHaveBeenCalled()
        expect(mockPreventDefault).not.toHaveBeenCalled()
        expect(mockClosePopup).not.toHaveBeenCalled()

        var submitButton = component.find(Button)
        expect(submitButton.text()).toBe("Edit")
        component.find(Form).simulate("submit", {preventDefault:mockPreventDefault})
        expect(store.dispatch).toHaveBeenCalled()
        expect(mockPreventDefault).toHaveBeenCalled()
        expect(mockClosePopup).toHaveBeenCalled()
    })
    it("submits add", () =>{
        initialState= {
            courses:[{
                "_id": "5db492c349a67b33b8d0a2a2",
                "course_name": "Calculus I",
                "course_description": "The basic of calculus (derivatives).",
                "start_term": "August 2019",
                "end_term": "December 2019",
                "number_of_lectures": 5,
                "instructor": "John Doe"
              }]
        }
        store = mockStore(initialState);
        store.dispatch = jest.fn()
        const mockPreventDefault = jest.fn();
        component = mount(<Provider store={store}><CourseForm closePopup={mockClosePopup}/></Provider>)

        expect(store.dispatch).not.toHaveBeenCalled()
        expect(mockPreventDefault).not.toHaveBeenCalled()
        
        var labels = ["courseName", "startDate", "endDate"]
        labels.forEach(l =>{
            var input = component.find("#"+l).at(0)
            input.simulate("change", {target:{value:"Calcu"}})
            component.update()
        })
        var submitButton = component.find(Button)
        expect(submitButton.text()).toBe("Tambahkan")
        component.find(Form).simulate("submit", {preventDefault:mockPreventDefault})
        expect(store.dispatch).toHaveBeenCalled()
        expect(mockPreventDefault).toHaveBeenCalled()
        component.update()
        var joinCourseAnnouncement = component.find(<JoinCourseAnnoucement closePopup={mockClosePopup}/>)
        expect(joinCourseAnnouncement).toBeDefined()
    })
})