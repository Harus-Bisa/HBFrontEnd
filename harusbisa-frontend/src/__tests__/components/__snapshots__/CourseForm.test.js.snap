// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CourseForm component matches snapshot 1`] = `"<div class=\\"container-fluid\\"><div class=\\"row\\" style=\\"border-bottom: 2px solid #FFE01C;\\"><div class=\\"col-12\\" style=\\"display: flex; justify-content: center; padding: 1rem 2rem;\\"><h3>Tambahkan Kelas Baru</h3></div></div><div class=\\"content\\"><form style=\\"min-width: 50vw;\\" class=\\"\\"><div class=\\"row form-group\\"><label class=\\"col-sm-4 col-form-label\\">Nama Kelas</label><div class=\\"col-sm-8\\"><input id=\\"courseName\\" placeholder=\\"Contoh: Biologi Molekuler Kelas A\\" type=\\"text\\" class=\\"form-control\\" value=\\"\\"><small class=\\"form-text text-muted\\">Nama ini adalah nama yang akan dilihat siswa Anda.</small></div></div><div class=\\"row form-group\\"><label class=\\"col-sm-4 col-form-label\\">Periode Kelas</label><div class=\\"col-sm-8\\"><div class=\\"row\\"><div class=\\"col-5\\"><select name=\\"select\\" id=\\"startDate\\" class=\\"form-control\\"><option>Juli 2019</option><option>Juli 2020</option><option>Juli 2021</option><option>Juli 2022</option><option>Juli 2023</option></select></div><div class=\\"col-2\\" style=\\"display: flex;\\"><p style=\\"margin: auto;\\">-</p></div><div class=\\"col-5\\"><select name=\\"select\\" id=\\"endDate\\" class=\\"form-control\\"><option>Juli 2019</option><option>Juli 2020</option><option>Juli 2021</option><option>Juli 2022</option><option>Juli 2023</option></select></div></div></div></div><div class=\\"row justify-content-end\\"><div class=\\"col-3\\"><button class=\\"MuiButtonBase-root MuiButton-root MuiButton-text prof-button Mui-disabled MuiButton-fullWidth Mui-disabled\\" tabindex=\\"-1\\" type=\\"submit\\" disabled=\\"\\"><span class=\\"MuiButton-label\\">Tambahkan</span></button></div></div></form></div></div>"`;
